File: 1723-maximum-number-of-achievable-transfer-requests/maximum-number-of-achievable-transfer-requests.cpp
class Solution {
public:
    int maxRequest = INT_MIN;

    void helper(vector<vector<int>>& requests, vector<int>& net, int idx, int n, int count) {
        // Base case
        if (idx == requests.size()) {
            bool allZeros = true;
            for (int i = 0; i < n; ++i) {
                if (net[i] != 0) {
                    allZeros = false;
                    break;
                }
            }

            if (allZeros) {
                maxRequest = max(maxRequest, count);
            }
            return;
        }

        int from = requests[idx][0];
        int to = requests[idx][1];

        // Take the current request
        net[from]--;
        net[to]++;
        helper(requests, net, idx + 1, n, count + 1);

        // Backtrack
        net[from]++;
        net[to]--;

        // Skip the current request
        helper(requests, net, idx + 1, n, count);
    }

    int maximumRequests(int n, vector<vector<int>>& requests) {
        vector<int> net(n, 0);
        helper(requests, net, 0, n, 0);
        return maxRequest;
    }
};

File: 1723-maximum-number-of-achievable-transfer-requests/maximum-number-of-achievable-transfer-requests.py
class Solution(object):
    def maximumRequests(self, n, requests):
        """
        :type n: int
        :type requests: List[List[int]]
        :rtype: int
        """
        self.maxRequest = float('-inf')
        
        def helper(net, idx, count):
            if idx == len(requests):
                if all(x == 0 for x in net):
                    self.maxRequest = max(self.maxRequest, count)
                return

            from_building, to_building = requests[idx]

            # Take the current request
            net[from_building] -= 1
            net[to_building] += 1
            helper(net, idx + 1, count + 1)

            # Backtrack
            net[from_building] += 1
            net[to_building] -= 1

            # Skip the current request
            helper(net, idx + 1, count)
        
        net = [0] * n
        helper(net, 0, 0)
        return self.maxRequest
